### DISCUSSION TITLE
I2C communication between a ESP32 and ESP32 Audiokit does not work

### DISCUSSION THREAD

**Discussion by Heute13** [pschatzmann/arduino-audio-driver#42] [Q&A] [answered]
Hello,
I want to send ranom values between 1 and 5 via I2C to my ESP32 Audiokit. Then I want to play a Audiofile, depending on which number arrives. The I2C communication works as long as i only want to print them on the serial monitor. As soon as I want to put in the code for the audio output it doesn't work anymore. The code for the audio output is the example "streams-sdmmc_wav-audiokit" in the arduino-audiotools-library. I think somhow i block the I2C used for the audio processing, when initializing the I2C between the two ESP.

I hope you can help me.
Thank you in advance

Errors that accured where for example: 
ES8388 not found on I2C address 10, check wiring, 
AudioBoard::driver::begin failed, 
setConfig has failed, 
AudioDriver init failed

The code for the audiokit I use is:

#include <Wire.h>
#include "FS.h"
#include "SD_MMC.h"
#include "AudioTools.h"
#include "AudioTools/AudioLibs/AudioBoardStream.h"

#define I2C_SLAVE_ADDR 0x08 // I²C-Adresse des Slave

volatile uint8_t sensor = 0;    // Zahl von 1 bis 5
volatile uint16_t distance = 0; // Zahl von 0 bis 400

AudioBoardStream i2s(AudioKitEs8388V1);
WAVDecoder wav;
EncodedAudioStream encoded(&i2s, &wav);
File audioFile;
StreamCopy copier(encoded, audioFile);

void setup() {
  Serial.begin(115200);
  Serial.println("I²C Master bereit");

  Wire1.begin(21, 22); 

  auto config = i2s.defaultConfig(TX_MODE);
  config.sd_active = false;
  i2s.begin(config);
  i2s.setVolume(0.25);

  if (!SD_MMC.begin("/sdcard", false)) {
    Serial.println("SD_MMC Fehler");
    while (true);
  }
}

void playAudioFile(uint8_t sensorValue) {
  String filename = "/sinus" + String(sensorValue) + ".wav";
  Serial.print("Spiele Datei: ");
  Serial.println(filename);

  if (audioFile) {
    audioFile.close();
  }

  audioFile = SD_MMC.open(filename.c_str());
  if (!audioFile) {
    Serial.println("Datei nicht gefunden");
    return;
  }

  encoded.begin();
  bool bPlaying = true;
  while (bPlaying) {
    if (copier.copy() != 0) {
      bPlaying = false;
    }
  }
  Serial.println("Abspielen beendet");
}

void loop() {
  uint8_t bytesReceived = Wire.requestFrom(I2C_SLAVE_ADDR, 3);

  if (bytesReceived == 3) {
    uint8_t buffer[3];
    for (int i = 0; i < bytesReceived; i++) {
      buffer[i] = Wire.read(); 
    }

 
    sensor = buffer[0];
    distance = buffer[1] | (buffer[2] << 8);

   
    Serial.print("Empfangener Sensorwert: ");
    Serial.println(sensor);
    Serial.print("Empfangener Distanzwert: ");
    Serial.println(distance);

   
    if (sensor >= 1 && sensor <= 5) {
      playAudioFile(sensor);
    }
  } else {
    Serial.println("Fehler beim Empfangen der Daten");
  }

  delay(1000); 
}

This is the o for the ESP32 I use:

#include <Wire.h>

#define I2C_SLAVE_ADDR 0x08  

uint8_t sensor = 1;       
uint16_t distance = 0;      

void setup() {
  Wire.begin(I2C_SLAVE_ADDR);        
  Wire.onRequest(requestEvent);      
  Serial.begin(115200);
  Serial.println("I2C Slave bereit");
}

void loop() {
  delay(1000); 
}

void requestEvent() {
  Serial.println("Master fordert Daten an!");

  sensor = (sensor % 5) + 1;        
  distance = (distance + 20) % 401;    

  uint8_t data[3];
  data[0] = sensor;                   // Zahl von 1 bis 5
  data[1] = distance & 0xFF;          // (LSB)
  data[2] = (distance >> 8) & 0xFF;   // (MSB)

  Wire.write(data, 3);                
  Serial.print("Gesendete Daten: Sensor = ");
  Serial.print(sensor);
  Serial.print(", Distance = ");
  Serial.println(distance);
}

This is the code for the Auiokit whih works but only with serial monitor output of the values

// Master Aktorik
#include <Wire.h>
#include "FS.h"
#include "SD_MMC.h"
#include "AudioTools.h"
#include "AudioTools/AudioLibs/AudioBoardStream.h"

#define I2C_SLAVE_ADDR 0x08 

volatile uint8_t sensor = 0;       
volatile uint16_t distance = 0;    

void setup() {
  Wire.begin(); 
  Serial.begin(115200);
  Serial.println("I2C Master bereit");
}

void loop() {
  uint8_t bytesReceived = Wire.requestFrom(I2C_SLAVE_ADDR, 3); 

  Serial.print("Bytes empfangen: ");
  Serial.println(bytesReceived);

  if (bytesReceived == 3) {
    uint8_t buffer[3];
    for (int i = 0; i < bytesReceived; i++) {
      buffer[i] = Wire.read(); 
    }

    sensor = buffer[0];

    distance = buffer[1] | (buffer[2] << 8);

    Serial.print("Empfangener Sensorwert: ");
    Serial.println(sensor);
    Serial.print("Empfangener Distanzwert: ");
    Serial.println(distance);
  } else {
    Serial.println("Fehler beim Empfangen der Daten");
  }

  delay(1000); 
}

**Comment by pschatzmann** [comment, marked as answer]
Be careful: for AudioKitEs8388V1 Wire is used with the pins 32 and 33 to communicate with the codec!
And pins 21 and 22 are used by the PA and LED

```C++
class PinsAudioKitEs8388v1Class : public DriverPins {
 public:
  PinsAudioKitEs8388v1Class() {
    // sd pins
    addSPI(ESP32PinsSD);
    // add i2c codec pins: scl, sda, port, frequency
    addI2C(PinFunction::CODEC, 32, 33);
    // add i2s pins: mclk, bck, ws,data_out, data_in ,(port)
    addI2S(PinFunction::CODEC, 0, 27, 25, 26, 35);

    // add other pins
    addPin(PinFunction::KEY, 36, PinLogic::InputActiveLow, 1);
    addPin(PinFunction::KEY, 13, PinLogic::InputActiveLow, 2);
    addPin(PinFunction::KEY, 19, PinLogic::InputActiveLow, 3);
    addPin(PinFunction::KEY, 23, PinLogic::InputActiveLow, 4);
    addPin(PinFunction::KEY, 18, PinLogic::InputActiveLow, 5);
    addPin(PinFunction::KEY, 5, PinLogic::InputActiveLow, 6);
    addPin(PinFunction::AUXIN_DETECT, 12, PinLogic::InputActiveLow);
    addPin(PinFunction::HEADPHONE_DETECT, 39, PinLogic::InputActiveLow);
    addPin(PinFunction::PA, 21, PinLogic::Output);
    addPin(PinFunction::LED, 22, PinLogic::Output);
  }
};

```

**Reply by Heute13** [reply]
Thank you for the hint. So where do I find two pins on the Audiokit I can use for I2C communication? And is it possible to use Wire1?

**Reply by Heute13** [reply]
I know thought of using Wire1.begin(5, 22);
because
 I do not need the key 8 and the LED. But I still get no connection. The Master sends no request to the slave and the slave no answer back.

**Reply by Heute13** [reply]
As soon as I remove i2s.begin(config); the I2C ommunication works. Could it be that I2S and I2C  interfere with each other?

**Reply by Heute13** [reply]
I have found a solution to my Problem. I initialize I2C, send the data, then I deinitialize I2C and initialize I2S. Then I want to open the audiofile on the sd card and play it, but the File can not be found. I double checked if the file name is right. 

Code:
#include <Wire.h>
#include "FS.h"
#include "SD_MMC.h"
#include "AudioTools.h"
#include "AudioTools/AudioLibs/AudioBoardStream.h"

// I²C-Konstanten
#define I2C_SLAVE_ADDR 0x55 // I²C-Adresse des Slave
#define sda 5
#define scl 18

// Array mit festen Dateinamen
const char *audioFiles[] = {
    "/sinus1.wav",  // Index 0 -> Sensor 1
    "/sinus2.wav",  // Index 1 -> Sensor 2
    "/sinus3.wav",  // Index 2 -> Sensor 3
    "/sinus4.wav",  // Index 3 -> Sensor 4
    "/sinus5.wav"   // Index 4 -> Sensor 5
};

// Variablen für empfangene Werte
volatile uint8_t sensor = 0;    // Zahl von 1 bis 5
volatile uint16_t distance = 0; // Zahl von 0 bis 400

// Audio-Objekte
AudioBoardStream i2s(AudioKitEs8388V1);
WAVDecoder wav;
EncodedAudioStream encoded(&i2s, &wav);
File audioFile;
StreamCopy copier(encoded, audioFile);

bool isI2SInitialized = false; // Zustand von I²S

// Funktion zum Initialisieren von I²C
void initI2C() {
  Wire1.begin(sda, scl); // I²C starten
  Serial.println("I²C initialisiert");
}

// Funktion zum Beenden von I²C
void deinitI2C() {
  Wire1.end(); // I²C stoppen
  Serial.println("I²C deaktiviert");
}

// Funktion zum Initialisieren von I²S
void initI2S() {
  if (!isI2SInitialized) {
    auto config = i2s.defaultConfig(TX_MODE);
    config.sd_active = false;
    i2s.begin(config);
    isI2SInitialized = true;
    Serial.println("I²S initialisiert");
  }
}

// Funktion zum Beenden von I²S
void deinitI2S() {
  if (isI2SInitialized) {
    i2s.end(); // I²S stoppen
    isI2SInitialized = false;
    Serial.println("I²S deaktiviert");
  }
}

// Funktion zum Abspielen einer Audiodatei basierend auf dem Sensorwert
void playAudioFile(uint8_t sensorValue) {
  if (sensorValue < 1 || sensorValue > 5) {
    Serial.println("Ungültiger Sensorwert für Audioauswahl");
    return;
  }

  // Wähle die Datei basierend auf dem Sensorwert
  const char *filename = audioFiles[sensorValue - 1]; // -1, da Array bei 0 beginnt
  Serial.print("Spiele Datei: ");
  Serial.println(filename);

  // Alte Datei schließen
  if (audioFile) {
    audioFile.close();
  }

  // Neue Datei öffnen
  audioFile = SD_MMC.open(filename);
  if (!audioFile) {
    Serial.println("Datei nicht gefunden");
    return;
  }

  // Dekoder-Stream starten
  encoded.begin();
  bool bPlaying = true;
  while (bPlaying) {
    if (copier.copy() != 0) {
      bPlaying = false;
    }
  }
  Serial.println("Abspielen beendet");
}

void setup() {
  // Serieller Monitor
  Serial.begin(115200);
  Serial.println("Systemstart");
  
  // SD-Karte einbinden
  if (!SD_MMC.begin("/sdcard", false)) {
    Serial.println("SD_MMC Fehler");
    while (true);
  }
}

void loop() {
  // **Phase 1: I²C aktivieren und Daten lesen**
  initI2C(); // I²C initialisieren
  delay(10); // Kurze Verzögerung zur Stabilisierung
  
  // Daten vom Slave abrufen
  uint8_t bytesReceived = Wire1.requestFrom(I2C_SLAVE_ADDR, 3);
  Serial.print("Bytes empfangen: ");
  Serial.println(bytesReceived);

  if (bytesReceived == 3) {
    uint8_t buffer[3];
    for (int i = 0; i < bytesReceived; i++) {
      buffer[i] = Wire1.read(); // Lese die empfangenen Bytes in das Array
    }

    // Sensor- und Distanzwerte aktualisieren
    sensor = buffer[0];
    distance = buffer[1] | (buffer[2] << 8);

    // Ausgabe der empfangenen Werte
    Serial.print("Empfangener Sensorwert: ");
    Serial.println(sensor);
    Serial.print("Empfangener Distanzwert: ");
    Serial.println(distance);
  } else {
    Serial.println("Fehler beim Empfangen der Daten");
  }

  deinitI2C(); // I²C deaktivieren

  // **Phase 2: I²S aktivieren und Audio abspielen**
  initI2S(); // I²S initialisieren
  delay(10); // Kurze Verzögerung zur Stabilisierung

  // Audio abspielen basierend auf Sensorwert
  playAudioFile(sensor);

  deinitI2S(); // I²S deaktivieren
  delay(1000); // Wartezeit zwischen den Durchläufen
}

**Comment by pschatzmann** [comment]
Did you try to separate the I2S from the I2C processing in separate tasks ?

**Reply by Heute13** [reply]
Thank you for your reply. 
Yes I did. But it does not work. The only way I got the I2C running, was as in the Code above. But therethe Problem accures that the Audiofile can not be found.

**Reply by Heute13** [reply]
The code does not work as soon as the I2S is initalized. Not when the I2S is used.

**Reply by pschatzmann** [reply]
Hmm, if there is no pin conflict, maybe the issue is the power supply ?

With SDMMC you are using 
   4   2   14   15 13  12  

With SD  you are using
  static PinsSPI ESP32PinsSD{PinFunction::SD, 14, 2, 15, 13, SPI};

**Reply by Heute13** [reply]
What do you mean? Why do I use SDMMC and SD? I am sorry if this is a dumb question, but I am new at this.

**Reply by Heute13** [reply]
I2S works always. I2C only when I2S is not initialzed.

**Reply by pschatzmann** [reply]
There are different SD APIs: the Arduino SD library uses SPI, SDMMC uses the ESP32 specific API

**Reply by Heute13** [reply]
Do I understand it right that I have to use sdmmc with the esp32 audiokit because it is everything permanently connected? WIth SD i would need the CS pin 5 which is not connected , am I right?

**Reply by pschatzmann** [reply]
No, you can also use the SD library which is included or the SDFAT library: You just need to set up the SPI with the correct pins.

This is something that this library does [automatically like in this example](https://github.com/pschatzmann/arduino-audio-tools/blob/main/examples/examples-audiokit/streams-sd_mp3-audiokit/streams-sd_mp3-audiokit.ino).
